packages/cli-kit/src/common/result.ts

```diff
@@ -0,0 +1,37 @@
+export type Result<T, TError = Error> = {ok: true; value: T} | {ok: false; error: TError}
+
+export type ResultAsync<T, TError = Error> = Promise<Result<T, TError>>
+
+export const ok = <T, TError = Error>(value: T): Result<T, TError> => {
+  return {ok: true, value}
+}
+
+const UnknownError = new Error('Unknown error')
+
+export const err = <T = never>(error?: unknown): Result<T, Error> => {
+  let errorToUse: Error
+  if (!error) {
+    errorToUse = UnknownError
+  } else if (error instanceof Error) {
+    errorToUse = error
+  } else if (typeof error === 'string') {
+    errorToUse = new Error(error)
+  } else {
+    errorToUse = UnknownError
+  }
+  return {ok: false, error: errorToUse}
+}
+
+export const valueOrThrow = <T, TError = Error>(result: Result<T, TError>, error?: Error): T => {
```

Comment:
This API opens the door for ignoring the error that's wrapped in the `result` parameter. What if we remove the second `error? parameter and have the following API instead:

```
mapError(result, (internalError) => { ... })
```

Reply:
But the `mapError` transform Result<T,E> to Result<T,U>. We could leave the API method `valueOrThrow` without the `error` parameter
At this point, why not using classes to define the result types and these API methods to that class. We will avoid having to pass the result as parameter and allows to concat several methods callings over the same result